File: coding/usaco/beads/beads.c
/*
ID: rohitya1
LANG: C
PROG: beads
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int mod (int a, int b) {
    return (a % b + b) % b;
}

int main() {
    FILE *fin = fopen("beads.in", "r");
    FILE *fout = fopen("beads.out", "w");

    int n;
    char beads[351];
    int idx, count;
    int max_beads = 0;
    int left_idx, right_idx;
    char left, right;

    fscanf(fin, "%d\n", &n);
    for (idx = 0; idx < n; idx++) fscanf(fin, "%c", &beads[idx]);

    for (idx = 0; idx < n; idx++) {
        right_idx = idx;
        left_idx = idx-1;
        right = beads[right_idx];
        left = beads[mod(left_idx, n)];

        count = 0;
        // count right
        for (; (beads[right_idx % n] == right || beads[right_idx % n] == 'w') && right_idx < (n+idx); right_idx++) count++;
        if (right == 'w') {
            right = beads[right_idx % n];
            for (; (beads[right_idx % n] == right || beads[right_idx % n] == 'w') && right_idx < (n+idx); right_idx++) count++;
        }

        // count left
        for (; (beads[mod(left_idx, n)] == left || beads[mod(left_idx, n)] == 'w') && left_idx >= -(n-idx) ; left_idx--) count++;
        if (left == 'w') {
            left = beads[mod(left_idx, n)];
            for (; (beads[mod(left_idx, n)] == left || beads[mod(left_idx, n)] == 'w') && left_idx >= -(n-idx) ; left_idx--) count++;
        }

        if (count > max_beads) max_beads = count;
        // possible all same colors
        if (max_beads >= n) {
            max_beads = n;
            break;
        }
    }

    fprintf(fout, "%d\n", max_beads);
    fclose(fin);
    fclose(fout);
    exit(0);
    return 0;
}

File: coding/usaco/beads/beads.java
/*
ID: rohitya1
LANG: JAVA
PROG: beads
*/

import java.util.*;
import java.io.*;
import java.lang.Math.*;


class beads {

    public static void main(String[] args) throws IOException {

        BufferedReader in = new BufferedReader(new FileReader("beads.in"));
        PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("beads.out")));

        int n = Integer.parseInt(in.readLine());
        String beads = in.readLine();
        beads = beads + beads; // 2 * beads string for linear traversal
        // Let's try dynamic programming
        int[][] left = new int[800][2];
        int[][] right = new int[800][2];
        left[0][0] = 0; //red
        left[0][1] = 0; //blue
        right[2*n][0] = 0;
        right[2*n][1] = 0;

        char color;
        for (int i = 1; i <= 2 * n; i++) {
            color = beads.charAt(i-1);
            if (color == 'r') {
                left[i][0] = left[i-1][0] + 1;
                left[i][1] = 0;
            } else if (color == 'b') {
                left[i][1] = left[i-1][1] + 1;
                left[i][0] = 0;
            } else {
                left[i][0] = left[i-1][0] + 1;
                left[i][1] = left[i-1][1] + 1;
            }

            int idx = 2*n - i;
            color = beads.charAt(idx);
            if (color == 'r') {
                right[idx][0] = right[idx+1][0] + 1;
                right[idx][1] = 0;
            } else if (color == 'b') {
                right[idx][1] = right[idx+1][1] + 1;
                right[idx][0] = 0;
            } else {
                right[idx][0] = right[idx+1][0] + 1;
                right[idx][1] = right[idx+1][1] + 1;
            }
        }

        int m = 0;
        for (int i = 0; i < 2*n; i++) {
            m = Math.max(m, Math.max(left[i][0], left[i][1]) + Math.max(right[i][0], right[i][1]));
        }
        out.println(Math.min(m, n));
        out.close();
        in.close();

        System.exit(0);
    }
}
